zilliqa/src/consensus.rs

```diff
@@ -1171,21 +1258,51 @@ impl Consensus {
             let account_nonce = self.state.get_account(tx.signer)?.nonce;
             transaction_pool.insert_transaction(tx, account_nonce);
         }
+
+        // Return the early proposal
         Ok(Some((proposal, broadcasted_transactions)))
     }
 
+    /// Produces the Proposal block.
+    /// It must return a final Proposal with correct QC, regardless of whether it is empty or not.
     fn propose_new_block(&mut self) -> Result<Option<(Block, Vec<VerifiedTransaction>)>> {
         let mut state = self.state.clone();
-        let mut tx_pool = self.transaction_pool.clone();
-        let mut votes = self.votes.clone();
 
-        let result = self.propose_new_block_at(&mut state, &mut tx_pool, &mut votes);
+        // give it a last try, even if it results in an empty block
+        if self.early_proposal.is_none() {
+            trace!("missing early proposal {}", self.view.get_view());
+            let mut tx_pool = self.transaction_pool.clone();
+            self.early_proposal =
+                self.assemble_early_block_at(&mut state, &mut tx_pool, &self.votes)?;
+            self.transaction_pool = tx_pool;
+        }
+        let (pending_block, applied_txs) = self.early_proposal.take().unwrap(); // safe to unwrap due to check above
+
+        // finalise the proposal
+        let final_block = self
+            .finish_early_proposal_at(&mut state, &self.votes, pending_block)?
+            .context("final block failure")?;
 
         self.state = state;
-        self.transaction_pool = tx_pool;
-        self.votes = votes;
 
-        result
+        Ok(Some((final_block, applied_txs)))
+    }
+
+    /// Assembles the Proposal block, but with a dummy QC.
+    pub fn assemble_early_block(&mut self) -> Result<()> {
+        // If no early_proposal exists, assemble with any ready transactions in pool.
+        // If the pool is empty, the early block assembly could be retried later.
+        // If this is called on each Vote, the Proposer can try several times to populate a block.
```

Comment:
> If this is called on each Vote, the Proposer can try several times to populate a block.

Does this mean that `assemble_early_block` is called when each vote arrives? Won't the second time do nothing because of the `if self.early_proposal.is_none()` condition below?

Reply:
If the first vote arrived and there was nothing in the pool, the second vote would still trigger the block assembly. But otherwise, you're correct. So, the Proposer can try to populate it several times, but will actually only do it once.